File: LeetCodeKotlin/AverageOfLevelsInBinaryTree_637.kt
class AverageOfLevelsInBinaryTree_637 {
    fun averageOfLevels(root: TreeNode?): DoubleArray {
        if (root == null) return DoubleArray(0)

        val result = mutableListOf<Double>()
        val queue = ArrayDeque<TreeNode>()
        queue.addLast(root)

        while (queue.isNotEmpty()) {
            val levelSize = queue.size
            var levelSum = 0L  // Use Long to avoid potential overflow when summing

            repeat(levelSize) {
                val currentNode = queue.removeFirst()
                levelSum += currentNode.value

                currentNode.left?.let { queue.addLast(it) }
                currentNode.right?.let { queue.addLast(it) }
            }

            result.add(levelSum.toDouble() / levelSize)
        }

        return result.toDoubleArray()
    }
}

File: LeetCodeCpp/problems/trees/AverageOfLevelsInBinaryTree_637.cpp
#include "AverageOfLevelsInBinaryTree_637.h"

#include <queue>

vector<double> AverageOfLevelsInBinaryTree_637::averageOfLevels(TreeNode<int>* root) {
    vector<double> result;

    if (root == nullptr) {
        return result;
    }

    queue<TreeNode<int>*> nodes;
    nodes.push(root);

    while (!nodes.empty()) {
        const size_t levelSize = nodes.size();
        long long sum = 0;

        for (size_t i = 0; i < levelSize; ++i) {
            TreeNode<int>* const node = nodes.front();
            nodes.pop();

            sum += node->val;

            if (node->left != nullptr) {
                nodes.push(node->left);
            }
            if (node->right != nullptr) {
                nodes.push(node->right);
            }
        }

        result.push_back(
            static_cast<double>(sum) / static_cast<double>(levelSize));
    }

    return result;
}
